### PR TITLE
MG-43: Implement RNA DE Aggregate ETL Transform

### PR DESCRIPTION
**Pull request by beatrizsaldana**
# [MG-43](https://sagebionetworks.jira.com/browse/MG-43) - Implement RNA DE Aggregate ETL Transform

## Problem

We need to generate a dataset to support the Model AD Explorer’s Gene Expression CT interface. This dataset needs to aggregate RNA differential expression data from multiple source files, with specific requirements for:

- Filtering human genes (ENSG*) and keeping only mouse genes (ENSMUSG*)
- Grouping data by gene, model, tissue, and sex
- Creating age-based entries with log2_fc and adj_p_val values
- Mapping gene metadata, biodomains, and model information
- Special handling for JAX models (tissue name mapping)
- Scientific notation formatting (5 significant digits)

### New Transform Module
- **`src/agoradatatools/etl/transform/rna_de_aggregate.py`** - Complete implementation of the RNA differential expression aggregation transform
- **`tests/transform/test_rna_de_aggregate.py`** - Comprehensive test suite with 528 lines of test coverage
- **Test assets** - Complete set of test data files in `tests/test_assets/rna_de_aggregate/`

### New File on Synapse to Track Datafiles
I created [rna_de_aggregate_data_files.csv](https://www.synapse.org/Synapse:syn69904462) to make it easier for us to track exactly which data files are being used for this transform and prevent us from having to make any repo changes to add or remove data files.

### Key Features Implemented

**Data Processing:**
- **Gene filtering**: Automatically filters out human genes (ENSG*) and keeps only mouse genes (ENSMUSG*)
- **Data aggregation**: Groups by gene, model, tissue, and sex with age-based entries containing log2_fc and adj_p_val
- **Memory optimization**: Processes files one at a time to reduce memory usage with garbage collection
- **Scientific notation handling**: Rounds log2_fc and adj_p_val to 5 significant digits (no scientific notation)

**Data Mapping & Transformations:**
- **Gene metadata lookup**: Maps ensembl_gene_id to gene_symbol using mouse_gene_metadata
- **Biodomain mapping**: Associates genes with their biodomains using biodom_genes_mm data
- **Model information**: Resolves model names, matched controls, and model types
- **JAX tissue mapping**: Special handling for JAX models - maps "Right Cerebral Hemisphere" to "Hemibrain"
- **Age sorting**: Numerically sorts age entries for consistent output

**Data Validation:**
- **Input validation**: Validates required datasets and columns
- **File validation**: Checks for empty files and missing required columns
- **Error handling**: Comprehensive error messages for debugging

### Output Structure

The transform generates JSON output with the following structure per gene/model/tissue/sex combination:

```json
[
  {
    "ensembl_gene_id": "ENSMUSG00000000001",
    "gene_symbol": "Gnai3",
    "biodomains": [
      "Apoptosis",
      "Autophagy",
      "Cell Cycle",
      "Metal Binding and Homeostasis",
      "Oxidative Stress",
      "Proteostasis",
      "Proteostasis",
      "Proteostasis",
      "Structural Stabilization",
      "Synapse",
      "Vasculature"
    ],
    "name": "5xFAD (Jax/IU/Pitt)",
    "matched_control": "C57BL/6J",
    "model_group": null,
    "model_type": "Familial AD",
    "tissue": "Hemibrain",
    "sex": "Females",
    "4 months": {
      "log2_fc": 0.01167,
      "adj_p_val": 0.7812
    },
    "12 months": {
      "log2_fc": 0.0055394,
      "adj_p_val": 0.94876
    }
  },
  {
    "ensembl_gene_id": "ENSMUSG00000000001",
    "gene_symbol": "Gnai3",
    "biodomains": [
      "Apoptosis",
      "Autophagy",
      "Cell Cycle",
      "Metal Binding and Homeostasis",
      "Oxidative Stress",
      "Proteostasis",
      "Proteostasis",
      "Proteostasis",
      "Structural Stabilization",
      "Synapse",
      "Vasculature"
    ],
    "name": "5xFAD (Jax/IU/Pitt)",
    "matched_control": "C57BL/6J",
    "model_group": null,
    "model_type": "Familial AD",
    "tissue": "Hemibrain",
    "sex": "Females & Males",
    "4 months": {
      "log2_fc": 0.0012218,
      "adj_p_val": 0.97786
    },
    "12 months": {
      "log2_fc": 0.0071723,
      "adj_p_val": 0.89033
    }
  },
...
```

### Performance Optimizations

- **Memory efficient**: Processes one file at a time instead of loading all files into memory
- **Dictionary lookups**: Pre-computed lookup dictionaries for O(1) data access
- **Garbage collection**: Explicit memory cleanup after processing each file

### Comprehensive Test Suite

**Test Coverage Includes:**
- **Happy path testing**: Valid data transformation with expected output
- **Error handling**: Missing datasets, empty files, missing columns
- **Gene filtering**: Human vs mouse gene filtering
- **JAX tissue mapping**: Special tissue name conversion
- **Age sorting**: Numerical age ordering
- **Data validation**: Input validation and error scenarios

**Test Files Created:**
- Input test data files (CSV format)
- Expected output JSON files
- Edge case test scenarios

**Test Classes:**
- `TestTransformRnaDeAggregate` - Main transform functionality tests
- `TestQuickValidateDataFile` - Data validation utility tests

### Test Scenarios

1. **Valid Data Transformation**
   - Tests complete data flow with expected output
   - Validates all mapping and aggregation logic

2. **Error Handling**
   - Missing required datasets
   - Empty data files
   - Missing required columns in data files

3. **Data Filtering**
   - Human gene filtering (ENSG* → filtered out)
   - Mouse gene retention (ENSMUSG* → kept)

4. **Special Cases**
   - JAX model tissue mapping ("Right Cerebral Hemisphere" → "Hemibrain")
   - Age entry numerical sorting
   - Biodomain association
   - Model information resolution

5. **Data Validation**
   - Input dataset validation
   - File structure validation
   - Column requirement validation

[MG-43]: https://sagebionetworks.jira.com/browse/MG-43?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/agoradatatools/etl/utils.py [hunk 1]
```diff
@@ -364,3 +364,31 @@ def convert_numpy_types(obj: Any) -> Any:
     elif isinstance(obj, list):
         return [convert_numpy_types(item) for item in obj]
     return obj
+
+
+def input_validation_model_info(df: pd.DataFrame) -> None:
```
Comment (beatrizsaldana) [inline src/agoradatatools/etl/utils.py:369]: I just moved this out of disease_correlation. No need to review this, it has already been reviewed and approved.

File: tests/test_utils.py [hunk 2]
```diff
@@ -726,3 +726,107 @@ def test_remove_duplicates_mixed_types_true_1(self):
     def test_remove_duplicates_preserves_order(self):
         input_list = ["a", "b", "a", "c", "b", "d"]
         assert utils.remove_duplicates_keep_order(input_list) == ["a", "b", "c", "d"]
+
```
Comment (beatrizsaldana) [inline tests/test_utils.py:729]: I just moved this from the disease_correlation testing file to the utilities testing file. No need to review this, it has already been approved.

File: tests/transform/test_disease_correlation.py [hunk 3]
```diff
@@ -618,107 +617,3 @@ def test_process_group_with_list_matched_controls(self):
 
         # Should take first element from the list
         assert result["matched_control"] == "C57BL6J"
-
```
Comment (beatrizsaldana) [inline tests/transform/test_disease_correlation.py:621]: Moved this to the utilities testing file. No need to review.

File: src/agoradatatools/etl/transform/disease_correlation.py [hunk 4]
```diff
@@ -68,34 +69,6 @@ def create_lookup(df: pd.DataFrame, group_by_col: str) -> Dict[str, Dict[str, An
     return lookup
 
 
-def input_validation_model_info(df: pd.DataFrame) -> None:
```
Comment (beatrizsaldana) [inline src/agoradatatools/etl/transform/disease_correlation.py:71]: Moved this to utils.py. No need to review this.

File: src/agoradatatools/etl/transform/disease_correlation.py
Comment (beatrizsaldana) [inline src/agoradatatools/etl/transform/disease_correlation.py:1]: No need to review changes to this file. Just moving a function from disease correlation to utils.

File: src/agoradatatools/etl/utils.py
Comment (beatrizsaldana) [inline src/agoradatatools/etl/utils.py:1]: No need to review changes to this file. Just moving a function from disease correlation to utils.

File: tests/test_utils.py
Comment (beatrizsaldana) [inline tests/test_utils.py:1]: No need to review changes to this file. Just moving a function from disease correlation to utils.

File: tests/transform/test_disease_correlation.py
Comment (beatrizsaldana) [inline tests/transform/test_disease_correlation.py:1]: No need to review changes to this file. Just moving a function from disease correlation to utils.

File: modelad_test_config.yaml [hunk 5]
```diff
@@ -140,3 +140,29 @@ datasets:
         gene: modified_gene
         ensembl_id: human_ensembl_id
         model: name
+
+  - rna_de_aggregate:
+      files:
+        - name: rna_de_aggregate_data_files
+          id: syn69904462
```
Comment (jaclynbeck-sage) [inline modelad_test_config.yaml]: This syn ID is missing from provenance below

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    # Process files one at a time to reduce memory usage
+    file_list = datasets["rna_de_aggregate_data_files"]
+    total_files = len(file_list)
+    syn = _login_to_synapse()
+
+    for i, (file_name, syn_id) in enumerate(file_list.itertuples(index=False)):
+        # Download and process one file at a time
+        data_file = get_entity_as_df(syn_id=syn_id, source="csv", syn=syn)
+        logger.info(
+            f"Processing {file_name} ({i+1}/{total_files}): {len(data_file)} rows, {len(data_file.columns)} columns, {data_file.memory_usage(deep=True).sum() / 1024**2:.2f} MB"
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: Break this line up for better readability:
```suggestion
            f"Processing {file_name} ({i+1}/{total_files}): {len(data_file)} rows, "
            f"{len(data_file.columns)} columns, "
            f"{data_file.memory_usage(deep=True).sum() / 1024**2:.2f} MB"
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            # Create age-based entries
+            age_entries = {}
+            for _, row in group.iterrows():
+                age = str(row["age"])
+                age_entries[age] = {
+                    "log2_fc": float(f"{float(row['log2FoldChange']):.5g}"),
+                    "adj_p_val": float(f"{float(row['padj']):.5g}"),
+                }
+
+            # If "name" includes "Jax" (case-insensitive) and tissue is "Right Cerebral Hemisphere", change tissue to "Hemibrain"
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: Make sonarcloud happy:
```suggestion
            # If "name" includes "Jax" (case-insensitive) and tissue is "Right Cerebral Hemisphere",
            # change tissue to "Hemibrain"
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    Transforms the rna_de_aggregate source files into a structured format for Model AD.
+    Groups by gene, model, tissue, and sex, with age-based entries containing log2_fc and adj_p_val.
+    """
+    check_required_datasets_and_columns(datasets, required_input)
+
+    # Pre-compute lookup dictionaries for efficient lookups
+    rnaseq_genotype_label_map_df = datasets["rnaseq_genotype_label_map"].fillna("")
+    mouse_gene_metadata_df = datasets["mouse_gene_metadata"].fillna("")
+    model_info_df = datasets["model_info"].fillna("")
+    biodom_genes_mm_df = datasets["biodom_genes_mm"].fillna("")
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: You should drop rows with `NA` Ensembl IDs before filling NA here, so there isn't a giant entry with key `""` in the lookup dict below. 
```suggestion
biodom_genes_mm_df = datasets["biodom_genes_mm"].dropna(axis = "index", subset = "ensembl_id").fillna("")
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            for _, row in group.iterrows():
+                age = str(row["age"])
+                age_entries[age] = {
+                    "log2_fc": float(f"{float(row['log2FoldChange']):.5g}"),
+                    "adj_p_val": float(f"{float(row['padj']):.5g}"),
+                }
+
+            # If "name" includes "Jax" (case-insensitive) and tissue is "Right Cerebral Hemisphere", change tissue to "Hemibrain"
+            if "jax" in str(name).lower() and tissue == "Right Cerebral Hemisphere":
+                tissue = "Hemibrain"
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: This code doesn't work for Jax.IU.Pitt_APOE4.Trem2.R47H because the display name is set to "LOAD1" instead of having Jax in it. You will have to add LOAD1 as a special case.

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            # Get model type using dictionary lookup
+            model_type = model_info_dict.get(model, "")
+
+            # Create age-based entries
+            age_entries = {}
+            for _, row in group.iterrows():
+                age = str(row["age"])
+                age_entries[age] = {
+                    "log2_fc": float(f"{float(row['log2FoldChange']):.5g}"),
+                    "adj_p_val": float(f"{float(row['padj']):.5g}"),
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: This rounding can be moved outside of the `for / enumerate(grouped)` loop and done in a single, more-readable step:
```
data_file = data_file.round(decimals = 5)
```
And this code can be changed to:
```
"log2_fc": float(row['log2FoldChange']),
"adj_p_val": float(row['padj']),
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    label_map_dict = {}
+    model_group_dict = {}
+    for _, row in rnaseq_genotype_label_map_df.iterrows():
+        key = (row["model"], row["genotype"])
+        label_map_dict[key] = row["display_label"]
+        model_group_dict[row["model"]] = row["model_group"]
+
+    # Create biodomain lookup dictionary
+    biodomain_dict = (
+        biodom_genes_mm_df.groupby("ensembl_id")["biodomain"].apply(list).to_dict()
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: There are some rows that have duplicate biodomain names for the same ensembl id. To fix this: 
```suggestion
        biodom_genes_mm_df[["ensembl_id", "biodomain"]].drop_duplicates().groupby("ensembl_id")["biodomain"].apply(list).to_dict()
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    ].to_dict()
+    model_info_dict = model_info_df.set_index("model")["model_type"].to_dict()
+
+    # Create label map dictionaries for efficient lookups
+    label_map_dict = {}
+    model_group_dict = {}
+    for _, row in rnaseq_genotype_label_map_df.iterrows():
+        key = (row["model"], row["genotype"])
+        label_map_dict[key] = row["display_label"]
+        model_group_dict[row["model"]] = row["model_group"]
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: You can shortcut this just like you did with the other dfs above:
```suggestion
    label_map_dict = rnaseq_genotype_label_map_df.set_index(["model", "genotype"])["display_label"].to_dict()
    model_group_dict = rnaseq_genotype_label_map_df.set_index(["model", "genotype"])["model_group"].to_dict()
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    output = []
+
+    # Process files one at a time to reduce memory usage
+    file_list = datasets["rna_de_aggregate_data_files"]
+    total_files = len(file_list)
+    syn = _login_to_synapse()
+
+    for i, (file_name, syn_id) in enumerate(file_list.itertuples(index=False)):
+        # Download and process one file at a time
+        data_file = get_entity_as_df(syn_id=syn_id, source="csv", syn=syn)
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: I'm a little worried about these files not going through the normal _process_ function to standardize values and column names, etc. I'm not sure how to get around the file size issue, do we care how much memory we are using @JessterB ? I think reading all the files in at once would be < 5 GB total.

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            f"Processing {file_name} ({i+1}/{total_files}): {len(data_file)} rows, {len(data_file.columns)} columns, {data_file.memory_usage(deep=True).sum() / 1024**2:.2f} MB"
+        )
+
+        _quick_validate_data_file(file_name, data_file)
+
+        # Filter out rows with human gene ensembl IDs (ENSG*), keep only mouse (ENSMUSG*)
+        data_file = data_file[data_file["ensembl_gene_id"].str.startswith("ENSMUSG")]
+
+        # Group by gene, model, tissue, and sex to create one entry per group
+        grouped = data_file.groupby(["ensembl_gene_id", "model", "tissue", "sex"])
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: Per my comment below, you need to add "case" and "control" to the grouping to avoid clobbering multiple case or control genotypes in studies that have them. 

```suggestion
        grouped = data_file.groupby(["ensembl_gene_id", "model", "tissue", "sex", "case", "control"])
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+
+        for i, ((ensembl_gene_id, model, tissue, sex), group) in enumerate(grouped):
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]:
```suggestion
        for i, ((ensembl_gene_id, model, tissue, sex, case, control), group) in enumerate(grouped):
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+            # Get gene metadata using dictionary lookup
+            gene_symbol = gene_metadata_dict.get(ensembl_gene_id, "")
+
+            # Get case and control from first row of group
+            case = group.iloc[0]["case"]
+            control = group.iloc[0]["control"]
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: These two lines aren't making a valid assumption, which also messes up the rest of your code below. For example, see the Jax.IU.Pitt_APOE4.Trem2.R47H file: There are 3 different "case" genotypes per gene/model/tissue/sex, with 2 different ages each. This causes your age_entries code to loop through 6 entries instead of 2, and it keeps overwriting the 4 month and 12 month entries until it gets to the end. You end up with a final entry with logfc/padj that belongs to the genotype "APOE4-KI_WT; Trem2-R47H_homozygous" because that was the last genotype in the group, but name is assigned to "APOE4-KI_homozygous; Trem2-R47H_homozygous" because that was the first genotype in the group.

This same issue exists for the "control" variable for studies like UCI_ABCA7, which have 2 different controls.

You will need to extend the grouping to include case and control as well, as suggested above.

```suggestion
```

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+                    "gene_symbol": gene_symbol,
+                    "biodomains": biodomains,
+                    "name": name,
+                    "matched_control": matched_control,
+                    "model_group": model_group if model_group != "" else None,
+                    "model_type": model_type,
+                    "tissue": tissue,
+                    "sex": sex,
+                    **dict(
+                        sorted(age_entries.items(), key=lambda x: int(x[0].split()[0]))
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: Not necessary for this PR, but I wonder if it would be good to add a `sort_dict` util function at some point, since I've seen you have to sort dicts by age before.

File: tests/transform/test_rna_de_aggregate.py [hunk 18]
```diff
@@ -0,0 +1,635 @@
+import os
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py:36]: I haven't reviewed this file yet, I will do so once some of the above changes are made because you will likely have to add new test cases to cover multiple "case" and "control" values for a single model. And check that LOAD1 tissue gets renamed properly.

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+    if data_file.empty:
+        raise ValueError("Data file is empty")
+
+    # Validate required columns
+    missing_columns = [col for col in required_columns if col not in data_file.columns]
+    if missing_columns:
+        raise ValueError(
+            f"Missing required columns in {file_name} dataset: {', '.join(missing_columns)}. "
+            f"Please ensure the {file_name} dataset contains all required columns: {', '.join(required_columns)}."
+        )
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: This looks like duplication of the main verification function. I'd suggest just using that instead of having this function. I think it would look like, instead of `_quick_validate_data_file(file_name, data_file)`:

```
if data_file.empty:
        raise ValueError("Data file is empty")
check_required_datasets_and_columns({filename: data_file}, {filename: <list_of_required_columns>})
```
which should have the same effect I think.

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
I found a few issues surrounding data sets that have multiple different case/control pairs, plus some cleanup/readability suggestions. Excellent job on the optimizations though, I tested speed of lookup dicts vs just using pandas data frame functions/merges, and you definitely sped things up a lot.

File: src/agoradatatools/etl/transform/rna_de_aggregate.py (hunk 7 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/rna_de_aggregate.py]: Thank you!

File: src/agoradatatools/etl/transform/rna_de_aggregate.py (hunk 13 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/rna_de_aggregate.py]: I talked to Jess about it and we decided to send them through the regular processing pipeline. The file size issue is not ideal, but it is not affecting performance too much if I only look through them one at a time.

File: src/agoradatatools/etl/transform/rna_de_aggregate.py (hunk 9 above)
Comment (JessterB) [reply src/agoradatatools/etl/transform/rna_de_aggregate.py]: Note that ONLY jax models will use "Right Cerebral Hemisphere" so you can apply this value transform every time you find that string

File: src/agoradatatools/etl/transform/rna_de_aggregate.py (hunk 17 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/rna_de_aggregate.py]: Created [this ticket](https://sagebionetworks.jira.com/browse/MG-460) to track future improvement.

File: src/agoradatatools/etl/transform/rna_de_aggregate.py (hunk 16 above)
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/rna_de_aggregate.py]: I believe this has been fixed. I am not going to mark it as resolved because I don't want us to lose track of this. I'll mark it as resolved when we all agree it is resolved.

File: tests/test_assets/rna_de_aggregate/README_synthetic_datasets.md [hunk 20]
```diff
@@ -0,0 +1,138 @@
+# Synthetic Datasets for transform_rna_de_aggregate()
```
Comment (jaclynbeck-sage) [inline tests/test_assets/rna_de_aggregate/README_synthetic_datasets.md:1]: Amazing!! Thank you for writing this!

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+
+        # Round numeric columns to 5 decimal places for consistency
+        data_file = data_file.round(decimals=5)
+
+        # Group by gene, model, tissue, and sex to create one entry per group
+        grouped = data_file.groupby(
+            ["ensembl_gene_id", "model", "tissue", "sex", "case", "control"]
+        )
+
+        for i, (
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py]: Nitpick: Even though it's not used in the loop, you should use something other than `i` here since you also use it in the parent loop.

File: src/agoradatatools/etl/transform/rna_de_aggregate.py [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+        data_file = data_file[data_file["ensembl_gene_id"].str.startswith("ENSMUSG")]
+
+        # Round numeric columns to 5 decimal places for consistency
+        data_file = data_file.round(decimals=5)
+
+        # Group by gene, model, tissue, and sex to create one entry per group
+        grouped = data_file.groupby(
+            ["ensembl_gene_id", "model", "tissue", "sex", "case", "control"]
+        )
+
```
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/rna_de_aggregate.py:212]: Nitpick: Can you add a comment here that says we are doing it this way rather than with pandas operations and merges, as a speed-up? Just in case we have to review future changes, we (I) know not to mess with it.

File: tests/transform/test_rna_de_aggregate.py [hunk 23]
```diff
@@ -0,0 +1,539 @@
...
+
+        # Sort output data by ensembl_gene_id for deterministic comparison
+        output_data_sorted = sorted(output_data, key=lambda x: x["ensembl_gene_id"])
+        expected_data_sorted = sorted(expected_data, key=lambda x: x["ensembl_gene_id"])
+
+        # Compare output with expected
+        assert output_data_sorted == expected_data_sorted
+
+    def test_transform_rna_de_aggregate_missing_required_dataset(self):
+        """Test that missing required datasets raise ValueError."""
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py:327]: Nitpick, optional change: For a really comprehensive test, you could change this function to use
```
@pytest.mark.parametrize("missing_file", [<list all 5 synthetic files>])

... inside function, which now takes (self, missing_file) ...
datasets = self._load_synthetic_test_data(list( set([<full list>]) - set([missing_file])))
```
so it tests for missing one of each file.

Comment (jaclynbeck-sage) [reply tests/transform/test_rna_de_aggregate.py:327]: On second thought, you probably don't need to do this since we're not trying to test the checker function itself, just that the checker function runs inside the transform.

File: tests/transform/test_rna_de_aggregate.py [hunk 24]
```diff
@@ -0,0 +1,539 @@
...
+        # Transform data
+        output_data = transform_rna_de_aggregate(datasets=datasets)
+
+        # Should only contain mouse genes (ENSMUSG*), not human genes (ENSG*)
+        assert len(output_data) == 2  # Only 2 mouse genes
+        for entry in output_data:
+            assert entry["ensembl_gene_id"].startswith("ENSMUSG")
+
+    def test_transform_rna_de_aggregate_jax_tissue_mapping(self):
+        """Test that JAX models with 'Right Cerebral Hemisphere' tissue are mapped to 'Hemibrain'."""
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py]: Is this something where you could just use the synthetic data and subset to jax model related rows? I see rows for Jax model in some of the files but I haven't poked at it to see if that would actually shorten this code or not.

File: tests/transform/test_rna_de_aggregate.py [hunk 25]
```diff
@@ -0,0 +1,539 @@
...
+        # Transform data
+        output_data = transform_rna_de_aggregate(datasets=datasets)
+
+        # Should map tissue from "Right Cerebral Hemisphere" to "Hemibrain"
+        assert len(output_data) == 1
+        assert output_data[0]["tissue"] == "Hemibrain"
+        assert "jax" in output_data[0]["name"].lower()
+
+    def test_transform_rna_de_aggregate_age_sorting(self):
+        """Test that age entries are sorted numerically."""
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py]: Same question w/ using synthetic data, but dropping in the age_data in place of something else.

File: tests/transform/test_rna_de_aggregate.py [hunk 26]
```diff
@@ -0,0 +1,539 @@
...
+        output_data = transform_rna_de_aggregate(datasets=datasets)
+
+        # Should have ages sorted numerically
+        assert len(output_data) == 1
+        entry = output_data[0]
+        age_keys = [
+            key
+            for key in entry.keys()
+            if key
+            not in [
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py]: You could clean this up a lot by:
```
[key for key in entry.keys() if "months" in key]
```
That _should_ work, I think...

File: tests/transform/test_rna_de_aggregate.py [hunk 27]
```diff
@@ -0,0 +1,539 @@
...
+        )
+        expected_data_sorted = sorted(
+            expected_data, key=lambda x: (x["ensembl_gene_id"], x["name"])
+        )
+
+        # Compare output with expected
+        assert output_data_sorted == expected_data_sorted
+
+    def test_synthetic_jax_tissue_mapping(self):
+        """Test JAX tissue mapping with synthetic data."""
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py]: Is this function fundamentally different from `test_transform_rna_de_aggregate_jax_tissue_mapping`?

File: tests/transform/test_rna_de_aggregate.py [hunk 28]
```diff
@@ -0,0 +1,539 @@
...
+                    df = df.rename(columns={"log2FoldChange": "log2foldchange"})
+
+                # Use the mapped key if available, otherwise use file name without extension
+                key = file_to_key_mapping.get(file_name, file_name.replace(".csv", ""))
+                datasets[key] = df
+
+        return datasets
+
+    def test_synthetic_basic_data(self):
+        """Test transformation with synthetic basic data."""
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py]: This function looks identical to `test_transform_rna_de_aggregate_should_pass`. A lot of these functions look identical actually.

File: tests/transform/test_rna_de_aggregate.py [hunk 29]
```diff
@@ -0,0 +1,539 @@
+import os
+import json
+
+import pandas as pd
+import pytest
+
+from agoradatatools.etl.transform.rna_de_aggregate import transform_rna_de_aggregate
+
+
+class TestTransformRnaDeAggregate:
```
Comment (jaclynbeck-sage) [inline tests/transform/test_rna_de_aggregate.py:291]: General comments on this file:

1. It looks like you have duplicates of most of the methods, named 2 different things. And several functions look like they _would_ be identical if synthetic files were dropped in in place of manually-created data. Please double-check and clean up duplicates.
2. Please add more helpful comments to each function's docstring, for functions where it's non-obvious what's in the file without going to look inside the file. For example, in `test_transform_rna_de_aggregate_should_pass`, briefly explain what "basic data" means. You don't need to explain empty data or missing file tests but things like "mixed genes" or "mixed models" needs a little more explanation on what part of the transform is actually being tested.

**Review by jaclynbeck-sage — CHANGES_REQUESTED** [review]
The main function looks great, I just have some nitpicks there. I do have some requested changes for the testing file, see my comments there.

File: tests/test_assets/rna_de_aggregate/README_synthetic_datasets.md (hunk 20 above)
Comment (JessterB) [reply tests/test_assets/rna_de_aggregate/README_synthetic_datasets.md:1]: +1!

File: tests/transform/test_rna_de_aggregate.py (hunk 27 above)
Comment (beatrizsaldana) [reply tests/transform/test_rna_de_aggregate.py]: I accidentally got a lot of duplicates after creating tests for synthetic data and not successfully deleting all of the pre-existing ones.

**Review by jaclynbeck-sage — APPROVED** [review]
Looks great! Amazing job, thank you for adding all the extra comments and doing such thorough error checking and testing.
